lib/src/dictionary.dart

```diff
@@ -0,0 +1,524 @@
+  /// have the value overwritten.
+  Dictionary<K, V> assoc(K key, V value);
+
+  /// Returns a new [Dictionary] with the keys and values from both.
+  ///
+  /// If key collisions occur, the values from the left Dictionary
+  /// are preserved.
+  Dictionary<K, V> merge(Dictionary<K, V> other);
+
+  /// Executes a function of [K] and [V] for each pair in [this].
+  void forEach(void f(K key, V value));
+
+  @override
+  String toString() {
+    final StringBuffer buffer = new StringBuffer();
+    buffer.write('(');
+    for (final pair in _pairs) {
+      buffer.write('${pair.first} -> ${pair.second}, ');
+    }
+    buffer.write(')');
+    return buffer.toString();
+  }
+}
+
+/// An internal only class for certain methods.
+class _Tuple2<K, V> {
+  final K first;
+  final V second;
+  const _Tuple2(this.first, this.second);
+}
+
+class _EmptyDictionary<K, V> extends Dictionary<K, V> {
+  static const String _rep = '()';
+
+  const _EmptyDictionary() : super._();
+
+  @override
+  Iterable<K> get keys => const [];
+
+  @override
+  Iterable<V> get values => const [];
+
+  @override
+  int get size => 0;
+
+  @override
+  bool containsKey(K _) => false;
+
+  @override
+  String toString() => _rep;
+
+  @override
+  Dictionary<K, V> remove(K _) => this;
+
+  @override
+  V operator [](K _) => null;
+
+  @override
+  Dictionary<K, V> merge(Dictionary<K, V> other) => other;
+
+  @override
+  Dictionary<K, V> assoc(K key, V value) {
+    final root = new BitmapIndexNode<K,V>.empty(0)
+      .assoc(0, key.hashCode, key, value);
+    return new _NodeDictionary._fromRoot(root, 1);
+  }
+
+  @override
+  Iterable<_Tuple2<K, V>> get _pairs => const [];
+
+  @override
+  void forEach(void f(K key, V value)) {}
+}
+
+class _NodeDictionary<K, V> extends Dictionary<K, V> {
+  final Node<K,V> _root;
+
+  @override
+  final int size;
+
+  _NodeDictionary._fromRoot(this._root, this.size): super._();
+
+  @override
+  Iterable<K> get keys => _root.traverseKeys();
+
+  @override
+  Iterable<V> get values => _root.traverseValues();
+
+
+  @override
+  bool containsKey(K key) => _root.find(key.hashCode, key) != null;
+
+  @override
+  Dictionary<K, V> remove(K key) {
+    final newNode = _root.remove(0, key.hashCode, key);
+    if (newNode != _root) {
+      return new _NodeDictionary._fromRoot(newNode, size - 1);
+    }
+    return this;
+  }
+
+  @override
+  V operator [](K key) {
+    return _root.find(key.hashCode, key);
+  }
+
+  @override
+  Dictionary<K, V> merge(Dictionary<K, V> other) => other;
+
+
+  @override
+  Dictionary<K, V> assoc(K key, V value) {
+    final newRoot = _root.assoc(0, key.hashCode, key, value);
+    return new _NodeDictionary._fromRoot(newRoot, size + 1);
+  }
+
+  @override
+  Iterable<_Tuple2<K, V>> get _pairs => _root.traverseBoth();
+
+  @override
+  void forEach(void f(K key, V value)) {
+    for (final tuple in _pairs) {
+      f(tuple.first, tuple.second);
+    }
+  }
+}
+
+/// bit magic
```

Comment:
They probably need more complete documentation, but I've at least added some one line descriptions of what they should be doing.